File: mobile/shared/src/commonMain/kotlin/com/sect/mobile/shared/presentation/viewmodel/PassationViewModel.kt
package com.sect.mobile.shared.presentation.viewmodel

import com.sect.mobile.shared.domain.repository.SECTRepositoryInterface
import com.sect.mobile.shared.presentation.action.PassationAction
import com.sect.mobile.shared.presentation.effect.AppEffect
import com.sect.mobile.shared.presentation.state.PassationState
import com.sect.mobile.shared.presentation.state.UiState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

class PassationViewModel(
    private val repository: SECTRepositoryInterface,
    private val scope: CoroutineScope
) {
    private val _state = MutableStateFlow(PassationState())
    val state: StateFlow<PassationState> = _state.asStateFlow()

    private val _effects = MutableSharedFlow<AppEffect>(extraBufferCapacity = 10)
    val effects = _effects.asSharedFlow()

    private var timerJob: kotlinx.coroutines.Job? = null
    private var autoSaveJob: kotlinx.coroutines.Job? = null
    private var currentSessionId: String? = null

    fun handleAction(action: PassationAction) {
        when (action) {
            is PassationAction.StartSession -> startSession(action.epreuveId)
            is PassationAction.OnReponseChanged -> onReponseChanged(action.questionId, action.contenu)
            PassationAction.NextQuestion -> nextQuestion()
            PassationAction.PreviousQuestion -> previousQuestion()
            PassationAction.SubmitSession -> submitSession()
            is PassationAction.GoToQuestion -> goToQuestion(action.index)
        }
    }

    private fun startSession(epreuveId: String) {
        scope.launch {
            _state.value = _state.value.copy(session = UiState.Loading)
            try {
                val session = repository.startSession(epreuveId)
                currentSessionId = session.id
                val epreuve = session.epreuve
                val questions = epreuve?.questions?.sortedBy { it.ordre } ?: emptyList()
                val existingReponses = session.reponses?.associate { it.questionId to (it.contenu ?: "") } ?: emptyMap()
                val totalSeconds = session.tempsRestant?.takeIf { it > 0 } ?: (epreuve?.duree?.times(60) ?: 0)

                _state.value = _state.value.copy(
                    session = UiState.Success(session),
                    epreuve = epreuve,
                    questions = questions,
                    localReponses = existingReponses,
                    remainingSeconds = totalSeconds,
                    isTimerRunning = true,
                    isSessionComplete = false
                )
                startTimer()
                startAutoSave()
            } catch (e: Exception) {
                _state.value = _state.value.copy(session = UiState.Error(e.message ?: "Erreur"))
            }
        }
    }

    private fun onReponseChanged(questionId: String, contenu: String) {
        _state.value = _state.value.copy(
            localReponses = _state.value.localReponses + (questionId to contenu)
        )
    }

    private fun nextQuestion() {
        val current = _state.value.currentQuestionIndex
        if (current < _state.value.questions.size - 1) {
            _state.value = _state.value.copy(currentQuestionIndex = current + 1)
        }
    }

    private fun previousQuestion() {
        val current = _state.value.currentQuestionIndex
        if (current > 0) {
            _state.value = _state.value.copy(currentQuestionIndex = current - 1)
        }
    }

    private fun goToQuestion(index: Int) {
        if (index in _state.value.questions.indices) {
            _state.value = _state.value.copy(currentQuestionIndex = index)
        }
    }

    private fun submitSession() {
        scope.launch {
            _state.value = _state.value.copy(isSubmitting = true)
            try {
                val sessionId = currentSessionId ?: return@launch
                val reponses = _state.value.localReponses.map { (qId, contenu) ->
                    mapOf("questionId" to qId, "contenu" to contenu)
                }
                val result = repository.submitSession(sessionId, reponses)
                _state.value = _state.value.copy(
                    session = UiState.Success(result),
                    isSubmitting = false,
                    isSessionComplete = true,
                    isTimerRunning = false
                )
                stopTimer()
                stopAutoSave()
                _effects.tryEmit(AppEffect.NavigateTo("results"))
            } catch (e: Exception) {
                _state.value = _state.value.copy(isSubmitting = false)
                _effects.tryEmit(AppEffect.ShowError(e.message ?: "Erreur de soumission"))
            }
        }
    }

    private fun startTimer() {
        timerJob = scope.launch {
            while (_state.value.remainingSeconds > 0) {
                delay(1000)
                val newSeconds = _state.value.remainingSeconds - 1
                _state.value = _state.value.copy(
                    remainingSeconds = newSeconds,
                    isTimeWarning = newSeconds < 300
                )
                if (newSeconds <= 0) {
                    _state.value = _state.value.copy(isTimerRunning = false)
                    submitSession()
                    break
                }
            }
        }
    }

    private fun stopTimer() {
        timerJob?.cancel()
        timerJob = null
    }

    private fun startAutoSave() {
        autoSaveJob = scope.launch {
            while (true) {
                delay(30_000)
                val sessionId = currentSessionId ?: break
                for ((qId, contenu) in _state.value.localReponses) {
                    try { repository.saveReponse(sessionId, qId, contenu) } catch (_: Exception) { }
                }
            }
        }
    }

    private fun stopAutoSave() {
        autoSaveJob?.cancel()
        autoSaveJob = null
    }

    fun formatTime(seconds: Int): String {
        val min = seconds / 60
        val sec = seconds % 60
        return String.format("%02d:%02d", min, sec)
    }

    fun cleanup() {
        stopTimer()
        stopAutoSave()
    }
}

File: mobile/iosApp/iosApp/ViewModels/PassationViewModel.swift
// SECT Mobile — iOS Passation ViewModel (Exam Session)
import SwiftUI
import Shared

@MainActor
class PassationViewModel: ObservableObject {
    @Published var session: SessionPassation? = nil
    @Published var epreuve: Epreuve? = nil
    @Published var questions: [Question] = []
    @Published var currentQuestionIndex: Int = 0
    @Published var answers: [String: String] = [:]  // questionId -> contenu
    @Published var remainingSeconds: Int = 0
    @Published var isTimerRunning = false
    @Published var isSubmitting = false
    @Published var isLoading = false
    @Published var error: String? = nil
    @Published var isSessionComplete = false

    private let repository = KoinRepositoryProvider.shared.repository
    private var timerTask: Task<Void, Never>? = nil
    private var autoSaveTask: Task<Void, Never>? = nil

    // ── Start Session ──

    func startSession(epreuveId: String) async {
        isLoading = true
        error = nil
        do {
            let session = try await repository.startSession(epreuveId: epreuveId)
            self.session = session
            self.epreuve = session.epreuve

            if let epreuve = session.epreuve, let qs = epreuve.questions {
                self.questions = qs.sorted { $0.ordre < $1.ordre }
            }

            // Load existing answers from session
            if let reponses = session.reponses {
                for reponse in reponses {
                    if let contenu = reponse.contenu {
                        answers[reponse.questionId] = contenu
                    }
                }
            }

            // Start countdown timer
            let totalSeconds: Int
            if let tempsRestant = session.tempsRestant, tempsRestant > 0 {
                totalSeconds = tempsRestant
            } else if let epreuve = self.epreuve {
                totalSeconds = epreuve.duree * 60
            } else {
                totalSeconds = 0
            }
            remainingSeconds = totalSeconds
            startTimer()
            startAutoSave()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    // ── Timer ──

    private func startTimer() {
        isTimerRunning = true
        timerTask = Task { [weak self] in
            while let self = self, self.remainingSeconds > 0, !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard !Task.isCancelled else { return }
                self.remainingSeconds -= 1
                if self.remainingSeconds <= 0 {
                    self.isTimerRunning = false
                    // Auto-submit when time runs out
                    await self.submitSession()
                }
            }
        }
    }

    func stopTimer() {
        timerTask?.cancel()
        timerTask = nil
        isTimerRunning = false
    }

    // ── Auto-Save ──

    private func startAutoSave() {
        autoSaveTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 30_000_000_000) // 30 seconds
                guard let self = self, !Task.isCancelled else { return }
                await self.saveAllAnswers()
            }
        }
    }

    func stopAutoSave() {
        autoSaveTask?.cancel()
        autoSaveTask = nil
    }

    // ── Question Navigation ──

    var currentQuestion: Question? {
        guard currentQuestionIndex >= 0, currentQuestionIndex < questions.count else { return nil }
        return questions[currentQuestionIndex]
    }

    var currentAnswer: String {
        get {
            guard let q = currentQuestion else { return "" }
            return answers[q.id] ?? ""
        }
        set {
            guard let q = currentQuestion else { return }
            answers[q.id] = newValue
        }
    }

    func goToQuestion(at index: Int) {
        guard index >= 0, index < questions.count else { return }
        currentQuestionIndex = index
    }

    func nextQuestion() {
        if currentQuestionIndex < questions.count - 1 {
            currentQuestionIndex += 1
        }
    }

    func previousQuestion() {
        if currentQuestionIndex > 0 {
            currentQuestionIndex -= 1
        }
    }

    var isFirstQuestion: Bool { currentQuestionIndex == 0 }
    var isLastQuestion: Bool { currentQuestionIndex == questions.count - 1 }

    // ── Answer Saving ──

    func saveCurrentAnswer() async {
        guard let session = session, let question = currentQuestion else { return }
        let contenu = answers[question.id] ?? ""
        do {
            try await repository.saveReponse(
                sessionId: session.id,
                questionId: question.id,
                contenu: contenu
            )
        } catch {
            // Silent fail for auto-save; user will see on submit
        }
    }

    func saveAllAnswers() async {
        guard let session = session else { return }
        for (questionId, contenu) in answers {
            do {
                try await repository.saveReponse(
                    sessionId: session.id,
                    questionId: questionId,
                    contenu: contenu
                )
            } catch { }
        }
    }

    // ── Submit Session ──

    func submitSession() async {
        guard let session = session else { return }
        stopTimer()
        stopAutoSave()
        isSubmitting = true
        error = nil
        do {
            // Save all answers first
            await saveAllAnswers()

            // Build reponses array for submission
            let reponses: [[String: Any?]] = answers.map { questionId, contenu in
                return [
                    "questionId": questionId,
                    "contenu": contenu
                ]
            }

            let result = try await repository.submitSession(
                sessionId: session.id,
                reponses: reponses
            )
            self.session = result
            isSessionComplete = true
        } catch {
            self.error = error.localizedDescription
        }
        isSubmitting = false
    }

    // ── Cleanup ──

    func cleanup() {
        stopTimer()
        stopAutoSave()
    }

    // ── Timer Display ──

    var timerDisplay: String {
        let minutes = remainingSeconds / 60
        let seconds = remainingSeconds % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    var timerProgress: Double {
        guard let epreuve = epreuve, epreuve.duree > 0 else { return 0 }
        let total = Double(epreuve.duree * 60)
        return Double(remainingSeconds) / total
    }

    var answeredCount: Int {
        answers.values.filter { !$0.isEmpty }.count
    }
}
